rclcpp/include/rclcpp/subscription_base.hpp

```diff
@@ -60,6 +63,14 @@ namespace experimental
 class IntraProcessManager;
 }  // namespace experimental
 
+enum class SubscriptionType : uint8_t
+{
+  ROS_MESSAGE = 1,  // take message as ROS message and handle as ROS message
+  SERIALIZED_MESSAGE = 2,  // take message as serialized and handle as serialized
+  DYNAMIC_MESSAGE_DIRECT = 3,  // take message as DynamicMessage and handle as DynamicMessage
+  DYNAMIC_MESSAGE_FROM_SERIALIZED = 4  // take message as serialized and handle as DynamicMessage
```

Comment:
The distinction here isn't  clear to me between dynamic message direct and serialized.

What does this wording pattern "take as X and handle as Y" mean? Does "handle" refer to the argument type for the callback?

Reply:
Yep, "handle" is for the callback.

The way it works with ROS is that a message has to be taken, then passed to a handle function that passes it to a callback. So take -> handle.

Reply:
What about a dynamic subscription that passes serialized data to the handler function? `GenericDynamicSubscription`, perhaps. It's what we would want in rosbag2 recording.

Reply:
In that case you would want to use a serialized subscription (I.e. one that takes and handles serialized data), then use the DynamicMessage class to deserialize to into the DynamicMessage

You probably would need to subclass that subscription so you can instantiate a DynamicMessage and have it accessible inside the callback

Reply:
```
    // Take serialized and then convert to dynamic message
    case rclcpp::SubscriptionType::DYNAMIC_MESSAGE_FROM_SERIALIZED:
      {
        std::shared_ptr<SerializedMessage> serialized_msg =
          subscription->create_serialized_message();

        // NOTE(methylDragon): Is this clone necessary? If I'm following the pattern, it seems so.
        DynamicMessage::SharedPtr dynamic_message = subscription->create_dynamic_message();
        take_and_do_error_handling(
          "taking a serialized message from topic",
          subscription->get_topic_name(),
          [&]() {return subscription->take_serialized(*serialized_msg.get(), message_info);},
          [&]()
          {
            // === <<< THIS IS THE MOST IMPORTANT LINE!!! >>> ===
            bool ret = dynamic_message->deserialize(&serialized_msg->get_rcl_serialized_message());
            if (!ret) {
              throw_from_rcl_error(ret, "Couldn't convert serialized message to dynamic data!");
            }
            subscription->handle_dynamic_message(dynamic_message, message_info);
          }
        );
        subscription->return_serialized_message(serialized_msg);
        subscription->return_dynamic_message(dynamic_message);
        break;
      }
  }
```